File: packages/sardis-sdk-python/src/sardis_sdk/resources/groups.py
"""
Groups resource for Sardis SDK.

This module provides both async and sync interfaces for managing agent groups.
Agent groups enable multi-agent governance with shared budgets and merchant policies.
"""
from __future__ import annotations

from typing import TYPE_CHECKING, Any, Dict, List, Optional, Union

from ..models.group import AgentGroup
from .base import AsyncBaseResource, SyncBaseResource

if TYPE_CHECKING:
    from ..client import AsyncSardisClient, SardisClient, TimeoutConfig


class AsyncGroupsResource(AsyncBaseResource):
    """Async resource for managing agent groups.

    Example:
        ```python
        async with AsyncSardisClient(api_key="...") as client:
            group = await client.groups.create(name="engineering-team")
            await client.groups.add_agent(group.group_id, "agent_abc")
        ```
    """

    async def create(
        self,
        name: str,
        budget: Optional[Dict[str, Any]] = None,
        merchant_policy: Optional[Dict[str, Any]] = None,
        metadata: Optional[Dict[str, Any]] = None,
        timeout: Optional[Union[float, "TimeoutConfig"]] = None,
    ) -> AgentGroup:
        """Create a new agent group."""
        payload: Dict[str, Any] = {"name": name}
        if budget is not None:
            payload["budget"] = budget
        if merchant_policy is not None:
            payload["merchant_policy"] = merchant_policy
        if metadata is not None:
            payload["metadata"] = metadata

        data = await self._post("groups", payload, timeout=timeout)
        return AgentGroup.model_validate(data)

    async def get(
        self,
        group_id: str,
        timeout: Optional[Union[float, "TimeoutConfig"]] = None,
    ) -> AgentGroup:
        """Get a group by ID."""
        data = await self._get(f"groups/{group_id}", timeout=timeout)
        return AgentGroup.model_validate(data)

    async def list(
        self,
        limit: int = 50,
        offset: Optional[int] = None,
        timeout: Optional[Union[float, "TimeoutConfig"]] = None,
    ) -> List[AgentGroup]:
        """List all agent groups."""
        params: Dict[str, Any] = {"limit": limit}
        if offset is not None:
            params["offset"] = offset

        data = await self._get("groups", params=params, timeout=timeout)
        if isinstance(data, list):
            return [AgentGroup.model_validate(item) for item in data]
        return [AgentGroup.model_validate(item) for item in data.get("groups", data.get("items", []))]

    async def update(
        self,
        group_id: str,
        name: Optional[str] = None,
        budget: Optional[Dict[str, Any]] = None,
        merchant_policy: Optional[Dict[str, Any]] = None,
        metadata: Optional[Dict[str, Any]] = None,
        timeout: Optional[Union[float, "TimeoutConfig"]] = None,
    ) -> AgentGroup:
        """Update a group."""
        payload: Dict[str, Any] = {}
        if name is not None:
            payload["name"] = name
        if budget is not None:
            payload["budget"] = budget
        if merchant_policy is not None:
            payload["merchant_policy"] = merchant_policy
        if metadata is not None:
            payload["metadata"] = metadata

        data = await self._patch(f"groups/{group_id}", payload, timeout=timeout)
        return AgentGroup.model_validate(data)

    async def delete(
        self,
        group_id: str,
        timeout: Optional[Union[float, "TimeoutConfig"]] = None,
    ) -> None:
        """Delete a group."""
        await self._delete(f"groups/{group_id}", timeout=timeout)

    async def add_agent(
        self,
        group_id: str,
        agent_id: str,
        timeout: Optional[Union[float, "TimeoutConfig"]] = None,
    ) -> AgentGroup:
        """Add an agent to a group."""
        data = await self._post(
            f"groups/{group_id}/agents",
            {"agent_id": agent_id},
            timeout=timeout,
        )
        return AgentGroup.model_validate(data)

    async def remove_agent(
        self,
        group_id: str,
        agent_id: str,
        timeout: Optional[Union[float, "TimeoutConfig"]] = None,
    ) -> AgentGroup:
        """Remove an agent from a group."""
        data = await self._delete(
            f"groups/{group_id}/agents/{agent_id}",
            timeout=timeout,
        )
        return AgentGroup.model_validate(data)

    async def get_spending(
        self,
        group_id: str,
        timeout: Optional[Union[float, "TimeoutConfig"]] = None,
    ) -> Dict[str, Any]:
        """Get current spending for a group."""
        return await self._get(f"groups/{group_id}/spending", timeout=timeout)


class GroupsResource(SyncBaseResource):
    """Sync resource for managing agent groups.

    Example:
        ```python
        with SardisClient(api_key="...") as client:
            group = client.groups.create(name="engineering-team")
            client.groups.add_agent(group.group_id, "agent_abc")
        ```
    """

    def create(
        self,
        name: str,
        budget: Optional[Dict[str, Any]] = None,
        merchant_policy: Optional[Dict[str, Any]] = None,
        metadata: Optional[Dict[str, Any]] = None,
        timeout: Optional[Union[float, "TimeoutConfig"]] = None,
    ) -> AgentGroup:
        """Create a new agent group."""
        payload: Dict[str, Any] = {"name": name}
        if budget is not None:
            payload["budget"] = budget
        if merchant_policy is not None:
            payload["merchant_policy"] = merchant_policy
        if metadata is not None:
            payload["metadata"] = metadata

        data = self._post("groups", payload, timeout=timeout)
        return AgentGroup.model_validate(data)

    def get(
        self,
        group_id: str,
        timeout: Optional[Union[float, "TimeoutConfig"]] = None,
    ) -> AgentGroup:
        """Get a group by ID."""
        data = self._get(f"groups/{group_id}", timeout=timeout)
        return AgentGroup.model_validate(data)

    def list(
        self,
        limit: int = 50,
        offset: Optional[int] = None,
        timeout: Optional[Union[float, "TimeoutConfig"]] = None,
    ) -> List[AgentGroup]:
        """List all agent groups."""
        params: Dict[str, Any] = {"limit": limit}
        if offset is not None:
            params["offset"] = offset

        data = self._get("groups", params=params, timeout=timeout)
        if isinstance(data, list):
            return [AgentGroup.model_validate(item) for item in data]
        return [AgentGroup.model_validate(item) for item in data.get("groups", data.get("items", []))]

    def update(
        self,
        group_id: str,
        name: Optional[str] = None,
        budget: Optional[Dict[str, Any]] = None,
        merchant_policy: Optional[Dict[str, Any]] = None,
        metadata: Optional[Dict[str, Any]] = None,
        timeout: Optional[Union[float, "TimeoutConfig"]] = None,
    ) -> AgentGroup:
        """Update a group."""
        payload: Dict[str, Any] = {}
        if name is not None:
            payload["name"] = name
        if budget is not None:
            payload["budget"] = budget
        if merchant_policy is not None:
            payload["merchant_policy"] = merchant_policy
        if metadata is not None:
            payload["metadata"] = metadata

        data = self._patch(f"groups/{group_id}", payload, timeout=timeout)
        return AgentGroup.model_validate(data)

    def delete(
        self,
        group_id: str,
        timeout: Optional[Union[float, "TimeoutConfig"]] = None,
    ) -> None:
        """Delete a group."""
        self._delete(f"groups/{group_id}", timeout=timeout)

    def add_agent(
        self,
        group_id: str,
        agent_id: str,
        timeout: Optional[Union[float, "TimeoutConfig"]] = None,
    ) -> AgentGroup:
        """Add an agent to a group."""
        data = self._post(
            f"groups/{group_id}/agents",
            {"agent_id": agent_id},
            timeout=timeout,
        )
        return AgentGroup.model_validate(data)

    def remove_agent(
        self,
        group_id: str,
        agent_id: str,
        timeout: Optional[Union[float, "TimeoutConfig"]] = None,
    ) -> AgentGroup:
        """Remove an agent from a group."""
        data = self._delete(
            f"groups/{group_id}/agents/{agent_id}",
            timeout=timeout,
        )
        return AgentGroup.model_validate(data)

    def get_spending(
        self,
        group_id: str,
        timeout: Optional[Union[float, "TimeoutConfig"]] = None,
    ) -> Dict[str, Any]:
        """Get current spending for a group."""
        return self._get(f"groups/{group_id}/spending", timeout=timeout)


__all__ = [
    "AsyncGroupsResource",
    "GroupsResource",
]

File: packages/sardis-sdk-js/src/resources/groups.ts
/**
 * Groups resource
 *
 * Agent groups enable multi-agent governance with shared budgets
 * and merchant policies across multiple agents.
 */

import { BaseResource } from './base.js';
import type {
  AgentGroup,
  CreateGroupInput,
  UpdateGroupInput,
  GroupSpending,
  RequestOptions,
} from '../types.js';

export class GroupsResource extends BaseResource {
  /**
   * Create a new agent group
   *
   * @example
   * ```typescript
   * const group = await client.groups.create({
   *   name: 'Engineering Team',
   *   budget: { daily: '5000.00', monthly: '50000.00' },
   * });
   * ```
   */
  async create(input: CreateGroupInput, options?: RequestOptions): Promise<AgentGroup> {
    return this._post<AgentGroup>('/api/v2/groups', input, options);
  }

  /**
   * Get a group by ID
   *
   * @example
   * ```typescript
   * const group = await client.groups.get('grp_abc123');
   * console.log(group.name, group.agent_ids);
   * ```
   */
  async get(groupId: string, options?: RequestOptions): Promise<AgentGroup> {
    return this._get<AgentGroup>(`/api/v2/groups/${groupId}`, undefined, options);
  }

  /**
   * List all agent groups
   *
   * @example
   * ```typescript
   * const groups = await client.groups.list();
   * const limited = await client.groups.list({ limit: 10 });
   * ```
   */
  async list(options?: { limit?: number; offset?: number }): Promise<AgentGroup[]> {
    const params: Record<string, unknown> = {};
    if (options?.limit !== undefined) params.limit = options.limit;
    if (options?.offset !== undefined) params.offset = options.offset;

    const response = await this._get<AgentGroup[] | { groups: AgentGroup[] }>(
      '/api/v2/groups',
      Object.keys(params).length > 0 ? params : undefined
    );

    if (Array.isArray(response)) return response;
    return response.groups || [];
  }

  /**
   * Update a group
   *
   * @example
   * ```typescript
   * const updated = await client.groups.update('grp_abc123', {
   *   name: 'Updated Name',
   *   budget: { daily: '10000.00' },
   * });
   * ```
   */
  async update(groupId: string, input: UpdateGroupInput): Promise<AgentGroup> {
    return this._patch<AgentGroup>(`/api/v2/groups/${groupId}`, input);
  }

  /**
   * Delete a group
   *
   * @example
   * ```typescript
   * await client.groups.delete('grp_abc123');
   * ```
   */
  async delete(groupId: string): Promise<void> {
    await this._delete(`/api/v2/groups/${groupId}`);
  }

  /**
   * Add an agent to a group
   *
   * @example
   * ```typescript
   * const group = await client.groups.addAgent('grp_abc123', 'agent_def456');
   * ```
   */
  async addAgent(groupId: string, agentId: string): Promise<AgentGroup> {
    return this._post<AgentGroup>(`/api/v2/groups/${groupId}/agents`, { agent_id: agentId });
  }

  /**
   * Remove an agent from a group
   *
   * @example
   * ```typescript
   * const group = await client.groups.removeAgent('grp_abc123', 'agent_def456');
   * ```
   */
  async removeAgent(groupId: string, agentId: string): Promise<AgentGroup> {
    return this._delete<AgentGroup>(`/api/v2/groups/${groupId}/agents/${agentId}`);
  }

  /**
   * Get current spending for a group
   *
   * @example
   * ```typescript
   * const spending = await client.groups.getSpending('grp_abc123');
   * console.log(spending.budget, spending.agent_count);
   * ```
   */
  async getSpending(groupId: string): Promise<GroupSpending> {
    return this._get<GroupSpending>(`/api/v2/groups/${groupId}/spending`);
  }
}
